### PR TITLE
feat: code editor console improvements

### PR DESCRIPTION
**Pull request by jimniels**
Generally enhances and re-styles the "console" in the editor.  Resolves part of #236 (still need to create formulas docs)

- Errors and `print()` log to the "Output" tab (errors are elevated)

CleanShot 2023-02-13 at 23 11 13@2x

- Language-specific context lives in the "About" tab

CleanShot 2023-02-13 at 23 12 59@2x

- Refactor language in the in the command palette to be shorter

CleanShot 2023-02-13 at 14 29 02@2x

- Console can be resized vertically

Feb-13-2023 23-16-31

### CODE REVIEW

File: src/constants/urls.ts [hunk 1]
```diff
@@ -1,3 +1,5 @@
 export const API_URL = 'http://localhost:8000';
 export const DOCUMENTATION_URL = 'https://docs.quadratichq.com';
+export const DOCUMENTATION_PYTHON_URL = `${DOCUMENTATION_URL}/reference/python-cell-reference`;
+export const DOCUMENTATION_FORMULAS_URL = `${DOCUMENTATION_URL}`; // TODO
```
Comment (jimniels) [inline src/constants/urls.ts:4]: Right now this just points to the docs home page as we don't have a formulas-specific page

**Comment by davidkircos** [comment]
All these changes look great to me. My one comment is changing the OUTPUT to be an input field so the user can select all inside of it and have a cursor be the default value, like VSCode.

Screenshot 2023-02-14 at 8 23 42 AM

**Comment by jimniels** [comment]
@davidkircos this is actually what it's doing now (maybe it wasn't for you?)

I didn't use an `input` or `textarea` because those don't allow you to style the text inside of them (and I think styling the error text is a nice touch)

CleanShot 2023-02-14 at 10 19 57@2x

It makes it feel more like VSCode which does the same:

CleanShot 2023-02-14 at 10 20 56@2x

So to achieve the same effect, [I use `contenteditable` and prevent editing the text](https://github.com/quadratichq/quadratic/pull/263/files#diff-2bc202dc21e9b27b3ad74739a0edda788d6d4da097d558b077a50d6fcc23276eR352-R363) this allows us to style the text of the output while still allowing the user to do `CMD` + `A` to select all text inside

Feb-14-2023 10-23-22

Also: this kind of implementation allows for a cursor to show up in the output box _when the user focuses it_

Feb-14-2023 10-24-38

This is similar to vscode, which won't show the cursor in the "output" tab unless the tab comes into focus by the user

Feb-14-2023 10-26-12

**Comment by jimniels** [comment]
@davidkircos maybe take one more look, as I refined the styles on the resize controls.

First, the resize control was a 1px line aligned to _one side_ of the target box.

CleanShot 2023-02-14 at 12 02 46@2x

Now it is a 1px line centered within with the target area

CleanShot 2023-02-14 at 11 59 47@2x

This makes the hover interactions feel more natural when you try to resize, as the target area is symmetrically balanced vs. asymmetrically balanced:

Feb-14-2023 11-46-13

Also: the resize target area now lights up when you hover over the area _with intent_ (e.g. slight pause over the area). VSCode has this when resizing:

Feb-14-2023 11-46-39

Which is a nice effect and now quadratic has it:

Feb-14-2023 11-46-34

All the above required CSS to achieve the effect, which is why you'll see it implemented as a CSS solution instead of inline styles via JS.

File: src/ui/menus/CodeEditor/CodeEditor.tsx [hunk 2]
```diff
@@ -196,164 +200,287 @@ export const CodeEditor = (props: CodeEditorProps) => {
...
+              }
+            }}
+            onCut={preventDefault}
+            onPaste={preventDefault}
+            style={{ outline: 'none' }}
+          >
+            {hasOutput && (
+              <>
+                {std_err && (
+                  <span style={{ display: 'flex', alignItems: 'center', gap: '8px', color: red[700] }}>
```
Comment (davidkircos) [inline src/ui/menus/CodeEditor/CodeEditor.tsx:369]: use `colors.error`

File: src/ui/menus/CodeEditor/CodeEditor.tsx [hunk 3]
```diff
@@ -196,164 +200,287 @@ export const CodeEditor = (props: CodeEditorProps) => {
...
+          <Console evalResult={evalResult} editorMode={editorMode} />
         )}
       </div>
-    );
-  return <></>;
+    </div>
+  );
 };
 
+function Console({ evalResult, editorMode }: { evalResult: cellEvaluationReturnType | undefined; editorMode: string }) {
```
Comment (davidkircos) [inline src/ui/menus/CodeEditor/CodeEditor.tsx:327]: move to a separate file

File: src/ui/menus/CodeEditor/CodeEditor.tsx [hunk 4]
```diff
@@ -196,164 +200,287 @@ export const CodeEditor = (props: CodeEditorProps) => {
...
+              </p>
+            </>
+          )}
+        </TabPanel>
+      </div>
+    </>
+  );
+}
+
+function LinkNewTab({ href, children }: { href: string; children: React.ReactNode }) {
```
Comment (davidkircos) [inline src/ui/menus/CodeEditor/CodeEditor.tsx:408]: Move to a new file. Can share a file with TabPanel

File: src/ui/menus/CodeEditor/CodeEditor.tsx [hunk 5]
```diff
@@ -196,164 +200,287 @@ export const CodeEditor = (props: CodeEditorProps) => {
...
+          }}
+        >
+          {children}
+        </pre>
+      )}
+    </div>
+  );
+}
+
+function ResizeControl({ setState, position }: { setState: Function; position: 'TOP' | 'LEFT' }) {
```
Comment (davidkircos) [inline src/ui/menus/CodeEditor/CodeEditor.tsx:444]: Move to new file

**Review by davidkircos — CHANGES_REQUESTED** [review]

File: src/ui/menus/CodeEditor/CodeEditor.tsx (hunk 4 above)
Comment (jimniels) [reply src/ui/menus/CodeEditor/CodeEditor.tsx]: Actually moved this to a new file and shared it with one very similar already used in the command palette, so now it works in both cases.

**Review by davidkircos — APPROVED** [review]

**Pull request merged**